### PR TITLE
Add new unittest for shared:recovery-key

### PR DESCRIPTION
**Pull request by devdekunle**
The unittest file contains tests cases for make_recovery_key, normalize_recovery_key and decode_modhex_char.

### CODE REVIEW

File: src/test/test-recovery-key.c [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        assert_se(decode_modhex_char('.') == -EINVAL);
+        assert_se(decode_modhex_char('/') == -EINVAL);
+        assert_se(decode_modhex_char('O') == -EINVAL);
+        assert_se(decode_modhex_char('\0') == -EINVAL);
+        assert_se(decode_modhex_char('J') == 8);
+
+}
+
+TEST(normalize_recovery_key) {
+        char *normalized_key = NULL;
```
Comment (bluca) [inline src/test/test-recovery-key.c]: needs _cleanup handling

File: src/test/test-recovery-key.c [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+        /* Case2: Normalization with dashes */
+        r = normalize_recovery_key("cdefVhij-cDefghij-cdefkhij-cdufghij-cdefgdij-cidefIhj-cdefNijR-cdVfguij",
+                        &normalized_key);
+
+        assert_se(r == 0);
+        assert_se(streq(normalized_key, "cdefvhij-cdefghij-cdefkhij-cdufghij-cdefgdij-cidefihj-cdefnijr-cdvfguij"));
+        free(normalized_key);
+
+        /* Case3: Invaled password length*/
```
Comment (bluca) [inline src/test/test-recovery-key.c]: missing space after "Case", same above

File: src/test/test-recovery-key.c [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+        /*Case 5: Normalization of Upper cases password without dashes*/
+        r = normalize_recovery_key("BFGHICEHHIUVLKJIHFHEDlntruvcdefjiTUVKLNIJVTUTKJIHDFBCBGHIJHHFDBC",
+                        &normalized_key);
+
+        assert(r == 0);
+        assert_se(streq(normalized_key, "bfghiceh-hiuvlkji-hfhedlnt-ruvcdefj-ituvklni-jvtutkji-hdfbcbgh-ijhhfdbc"));
+        free(normalized_key);
+
+        /*Case 5: Minimun password length */
```
Comment (bluca) [inline src/test/test-recovery-key.c]: this is 6, not 5

File: src/test/test-recovery-key.c [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        assert_se(decode_modhex_char('v') == 15);
+        assert_se(decode_modhex_char('V') == 15);
+        assert_se(decode_modhex_char('a') == -EINVAL);
+        assert_se(decode_modhex_char('A') == -EINVAL);
+        assert_se(decode_modhex_char('x') == -EINVAL);
+        assert_se(decode_modhex_char('.') == -EINVAL);
+        assert_se(decode_modhex_char('/') == -EINVAL);
+        assert_se(decode_modhex_char('O') == -EINVAL);
+        assert_se(decode_modhex_char('\0') == -EINVAL);
+        assert_se(decode_modhex_char('J') == 8);
```
Comment (bluca) [inline src/test/test-recovery-key.c]: duplicated

File: src/test/test-recovery-key.c [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                for (size_t i = 0; i < length; i++) {
+
+                        assert_se((generated_keys[test][i] >= 'a' && generated_keys[test][i] <= 'v')
+                                || generated_keys[test][i] == '-');
+
+                        if (i % 9 == 8)
+                                assert_se(generated_keys[test][i] == '-');
+                }
+                /* check for uniqueness of each generated recovery key*/
+                for (int prev = 0; prev < test; ++prev) {
```
Comment (bluca) [inline src/test/test-recovery-key.c]: no bracket around single line for

File: src/test/test-recovery-key.c [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+                        assert_se(!streq(generated_keys[test], generated_keys[prev]));
+                }
+
+        }
+        for (int i = 0; i < num_test; i++) {
```
Comment (bluca) [inline src/test/test-recovery-key.c]: use size_t, not int

File: src/test/test-recovery-key.c [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+                        assert_se((generated_keys[test][i] >= 'a' && generated_keys[test][i] <= 'v')
+                                || generated_keys[test][i] == '-');
+
+                        if (i % 9 == 8)
+                                assert_se(generated_keys[test][i] == '-');
+                }
+                /* check for uniqueness of each generated recovery key*/
+                for (int prev = 0; prev < test; ++prev) {
+
```
Comment (bluca) [inline src/test/test-recovery-key.c]: remove this empty line

File: src/test/test-recovery-key.c [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        for (int test = 0; test < num_test; ++test) {
+                r = make_recovery_key(&generated_keys[test]);
+
+                assert_se(r == 0);
+                length = strlen(generated_keys[test]);
+
+                assert_se(length == RECOVERY_KEY_MODHEX_FORMATTED_LENGTH - 1);
+
+                for (size_t i = 0; i < length; i++) {
+
```
Comment (bluca) [inline src/test/test-recovery-key.c]: remove this empty line

File: src/test/test-recovery-key.c [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        r = make_recovery_key(&recovery_key);
+
+        assert_se(r == 0);
+        assert_se(recovery_key != NULL);
+        /* Check that length of formatted key is 72 with 64 modhex characters*/
+        length = strlen(recovery_key);
+
+        assert_se(length == RECOVERY_KEY_MODHEX_FORMATTED_LENGTH - 1);
+        /* Check modhex characters in formatted key with dashes */
+        for(size_t i = 0; i < length; i++) {
```
Comment (bluca) [inline src/test/test-recovery-key.c]: missing space after for

File: src/test/test-recovery-key.c [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
```
Comment (bluca) [inline src/test/test-recovery-key.c]: remove this empty line

File: src/test/test-recovery-key.c [hunk 11]
```diff
@@ -0,0 +1,158 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <string.h>
+
+#include "random-util.h"
+#include "recovery-key.h"
+#include "memory-util.h"
```
Comment (bluca) [inline src/test/test-recovery-key.c]: sort alphabetically

File: src/test/test-recovery-key.c [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+        /* Check for succesful recovery-key creation */
+        r = make_recovery_key(&recovery_key);
+
+        assert_se(r == 0);
+        assert_se(recovery_key != NULL);
+        /* Check that length of formatted key is 72 with 64 modhex characters*/
+        length = strlen(recovery_key);
+
+        assert_se(length == RECOVERY_KEY_MODHEX_FORMATTED_LENGTH - 1);
```
Comment (bluca) [inline src/test/test-recovery-key.c:24]: move the assert on the next line of the things they check, rather than leaving empty lines, here and elsewhere

**Comment by devdekunle** [comment]
Thanks for your review @bluca, I have implemented the changes. Kindly re-review

File: src/test/test-recovery-key.c [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+        r = normalize_recovery_key("1234-5678-90AB-CDEF-1234-5678-90AB-CDEF", &normalized_key1);
+        assert(r == -EINVAL);
+
+
+        /* Case 4: Invalid password format(missing dash)*/
+        r = normalize_recovery_key("cdefghij-cdefghij-cdefghij-cdefghij-cdefghij-cdefghij-cdefghijcdefghij",
+                        &normalized_key1);
+        assert_se(r == -EINVAL);
+
+        /*Case 5: Normalization of Upper cases password without dashes*/
```
Comment (bluca) [inline src/test/test-recovery-key.c]: missing space before "Case" here and below

File: src/test/test-recovery-key.c [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+        /* Case 2: Normalization with dashes */
+        r = normalize_recovery_key("cdefVhij-cDefghij-cdefkhij-cdufghij-cdefgdij-cidefIhj-cdefNijR-cdVfguij",
+                        &normalized_key2);
+        assert_se(r == 0);
+        assert_se(streq(normalized_key2, "cdefvhij-cdefghij-cdefkhij-cdufghij-cdefgdij-cidefihj-cdefnijr-cdvfguij"));
+
+        /* Case 3: Invalid password length*/
+        r = normalize_recovery_key("1234-5678-90AB-CDEF-1234-5678-90AB-CDEF", &normalized_key1);
+        assert(r == -EINVAL);
+
```
Comment (bluca) [inline src/test/test-recovery-key.c:115]: extra new line

File: src/test/test-recovery-key.c [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+TEST(normalize_recovery_key) {
+        _cleanup_(erase_and_freep) char *normalized_key1 = NULL;
+        _cleanup_(erase_and_freep) char *normalized_key2 = NULL;
+        _cleanup_(erase_and_freep) char *normalized_key3 = NULL;
+        int r;
+
+        /* Case 1: Normalization without dashes */
+        r = normalize_recovery_key("cdefghijcdefghijcdefghijcdefghijcdefghijcdefghijcdefghijcdefghij",
+                        &normalized_key1);
+
```
Comment (bluca) [inline src/test/test-recovery-key.c]: extra newline

File: src/test/test-recovery-key.c [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+                assert_se((recovery_key[i] >= 'a' && recovery_key[i] <= 'v') || recovery_key[i] == '-');
+                if (i % 9 == 8)
+                        /* confirm '-' is after every 8 characters */
+                        assert_se(recovery_key[i] == '-');
+        }
+
+        /* Repeat tests to determine randomness */
+        for (size_t test = 0; test < num_test; ++test) {
+                r = make_recovery_key(&generated_keys[test]);
+
```
Comment (bluca) [inline src/test/test-recovery-key.c]: extra newline

File: src/test/test-recovery-key.c [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+                assert_se(r == 0);
+                length = strlen(generated_keys[test]);
+
+                assert_se(length == RECOVERY_KEY_MODHEX_FORMATTED_LENGTH - 1);
+
+                for (size_t i = 0; i < length; i++) {
+
```
Comment (bluca) [inline src/test/test-recovery-key.c]: extra newline

**Review by bluca — APPROVED** [review]

**Pull request merged**